ckanext/harvest/plugin/__init__.py

```diff
@@ -125,59 +155,37 @@ def before_dataset_index(self, pkg_dict):
-
-                # If the harvest extras are there, update them. This can
-                # happen eg when calling package_update or resource_update,
-                # which call package_show
-                harvest_not_found = True
-                harvest_not_found_validated = True
-                if not data_dict.get("extras"):
-                    data_dict["extras"] = []
-
-                for e in data_dict.get("extras"):
-                    if e.get("key") == key:
-                        e.update({"value": value})
-                        harvest_not_found = False
-                if harvest_not_found:
-                    data_dict["extras"].append({"key": key, "value": value})
-
-                if not validated_data_dict.get("extras"):
-                    validated_data_dict["extras"] = []
-
-                for e in validated_data_dict.get("extras"):
-                    if e.get("key") == key:
-                        e.update({"value": value})
-                        harvest_not_found_validated = False
-                if harvest_not_found_validated:
-                    validated_data_dict["extras"].append({"key": key, "value": value})
-
-                # The commented line isn't cataloged correctly, if we pass the
-                # basic key the extras are prepended and the system works as
-                # expected.
-                # pkg_dict['extras_{0}'.format(key)] = value
-                pkg_dict[key] = value
-
-            pkg_dict["data_dict"] = json.dumps(data_dict)
-            pkg_dict["validated_data_dict"] = json.dumps(validated_data_dict)
-
-        if isinstance(pkg_dict.get('status'), dict):
-            try:
-                pkg_dict['status'] = json.dumps(pkg_dict['status'])
-            except ValueError:
-                pkg_dict.pop('status', None)
+        if not harvest_object:
+            return pkg_dict
+
+        harvest_extras = [
+            ("harvest_object_id", harvest_object.id),
+            ("harvest_source_id", harvest_object.source.id),
+            ("harvest_source_title", harvest_object.source.title),
+        ]
+
+        # Add harvest extras to data_dict
+        data_dict = json.loads(pkg_dict["data_dict"])
+        for key, value in harvest_extras:
+            if key in data_dict.keys():
+                data_dict[key] = value
+                continue
+            self._add_or_update_harvest_metadata(key, value, data_dict)
+
+        # Add harvest extras to validated_data_dict
+        validated_data_dict = json.loads(pkg_dict["validated_data_dict"])
+        for key, value in harvest_extras:
+            if key in validated_data_dict.keys():
+                validated_data_dict[key] = value
+                continue
+            self._add_or_update_harvest_metadata(key, value, validated_data_dict)
+
+        # Add harvest extras to main indexed pkg_dict
+        for key, value in harvest_extras:
+            pkg_dict[key] = value
```

Comment:
This code has been here but I don't see it working. 

Indexed packages do not have this attributes when calling Solr directly.

Reply:
This was done to add the harvest field as the ["catch-all" string field](https://github.com/ckan/ckan/blob/98d2164f48efff6d2448acf3ba5dc0c8c0796fdc/ckan/config/solr/schema.xml#L175) in Solr. These values are indexed but not stored, so that's why you don't see them when calling Solr directly.
With your changes, we should only add them if not already there in `pkg_dict`